shell/artifacts/hex/hex-board.js

```diff
@@ -0,0 +1,120 @@
+      grid-template-rows: repeat(${size}, calc(0.75 * var(--hex-height) - var(--hex-border)));
+    }
+    
+    hexa-gon[offset] {
+      margin-left: calc(1 * var(--hex-x-offset) - var(--hex-border)/2);
+    }
+
+    hexa-gon[padding] {
+      visibility: hidden;
+    }
+    
+    /* Clipped hexagon tile. */
+    hexa-gon {
+      position: relative;
+      display: inline-block;
+      width: var(--hex-width);
+      height: var(--hex-height);
+      clip-path: polygon(
+          0% 25%,
+          50% 0%,
+          100% 25%,
+          100% 75%,
+          50% 100%,
+          0% 75%);
+    }
+
+    /* Border on top of the tile. */
+    hexa-gon:before {
+      position: absolute;
+      content: '';
+      width: var(--hex-width);
+      height: var(--hex-height);
+      --0: calc(var(--hex-border) + 0%);
+      --25: calc(var(--hex-border) + (100% - 2 * var(--hex-border)) * 0.25);
+      --75: calc(var(--hex-border) + (100% - 2 * var(--hex-border)) * 0.75);
+      --100: calc(100% - var(--hex-border));
+      clip-path: polygon(
+        0% 25%,
+        50% 0%,
+        100% 25%,
+        100% 75%,
+        50% 100%,
+        0% 75%,
+        0% 25%,
+        var(--0) var(--25),
+        var(--0) var(--75),
+        50% var(--100),
+        var(--100) var(--75),
+        var(--100) var(--25),
+        50% var(--0),
+        var(--0) var(--25));
+      background: black;
+    }
+    
+    hexa-gon:hover {
+      background: #F44336;
+    }
+    </style>
+    <hex-board>
+      ${(() => {
+        let result = [];
+        for (let row = 0; row < size; row++) {
+          // Every second row is offset by half the width of a tile (by setting
+          // the [offset] attribute). After every two rows we need to add an
+          // additional padding element. We must also add padding elements at
+          // the end so that each row has the same number of elements.
+          // eg. (Xx is a hex tile, . is the offset applied to every second
+          //     row, Pp is padding)
+          // XxXxXxXxXxXxPpPp
+          // .XxXxXxXxXxXxPpPp
+          // PpXxXxXxXxXxXxPp
+          // .PpXxXxXxXxXxXxPp
+          // PpPpXxXxXxXxXxXx
+          // .PpPpXxXxXxXxXxXx
+          let offset = row % 2;
+          for (let i = 0; i < (row - offset) / 2; i++) {
+            result.push(`<hexa-gon padding${offset ? ' offset' : ''}></hexa-gon>`);
+          }
+          for (let col = 0; col < size; col++) {
+            result.push(`<hexa-gon${offset ? ' offset' : ''}></hexa-gon>`);
+          }
+          for (let i = (row - offset) / 2; i < padding; i++) {
+            result.push(`<hexa-gon padding${offset ? ' offset' : ''}></hexa-gon>`);
+          }
+        }
+        return result.join('');
+      })()}
+    </hex-board>`;
+
+  return class extends DomParticle {
+    get template() {
+      return template;
+    }
+    update(props, state, lastProps, lastState) {
+    }
+    render(props, state) {
```

Comment:
Usually you only need `render`, it used to be that `update` was the only way to get at `lastProps` and `lastState` but @sjmiles changed that recently so render has it too.